refactor(transfer): align with select & monorepo structure

Mohammer5 opened the pull request:
Fixes #50

This commit does several things:
* Use strings as selected values
* Use strings to store which options are highlighted
* Removes jest test abstractions
* Simplifies internal names
* Replaces `common.js` files with files whose names match their content
* Moves the Transfer component from `core` to `widgets`
* Adds the `.js` extension to all import statements where applicable
* Provides the correct `disabled` & `dataTest` props to the re-order up
button's icon

BREAKING CHANGE: The Transfer component now expects options to be passed in as objects, not as children. Custom components can be provided via the `optionComponent` prop for all options or via the `component` property on an individual option.

BREAKING CHANGE: The Transfer component now expects strings as selected
values instead of option objects.

BREAKING CHANGE: The Transfer component is now part of `widgets`

Mohammer5 commented:
![image](https://user-images.githubusercontent.com/10101831/81277366-9b935c80-9054-11ea-9ee2-5ac77bbf4dba.png)

:+1:

Mohammer5 commented:
@HendrikThePendric The multi select feature failed, it does not fail locally. I assume that's another issue with the `requestAnimationFrame` timing, right?

ismay commented:
> @HendrikThePendric The multi select feature failed, it does not fail locally. I assume that's another issue with the `requestAnimationFrame` timing, right?

Yeah I think we need to take another at the select. I think we can find another way to deal with this than using RAF, and with the recent refactors we've had of the select I suspect there are bits of it that need a little love.

HendrikThePendric commented:
> The multi select feature failed, it does not fail locally. I assume that's another issue with the requestAnimationFrame timing, right?

Oh man 🤦 that's very unfortunate. It's not really an issue with `requestAnimationFrame`, but rather with the `ResizeObserver`. I had already implemented a helper called `waitForResizeObserver` to fix it, but apparently that isn't super robust either. I really don't know of another way to address this problem.

ismay commented:
> > The multi select feature failed, it does not fail locally. I assume that's another issue with the requestAnimationFrame timing, right?
> 
> Oh man 🤦 that's very unfortunate. It's not really an issue with `requestAnimationFrame`, but rather with the `ResizeObserver`. I had already implemented a helper called `waitForResizeObserver` to fix it, but apparently that isn't super robust either. I really don't know of another way to address this problem.

It's weird, I tried rerunning the integration tests, but now they're failing completely:

<img width="875" alt="Screenshot 2020-05-11 at 11 09 52" src="https://user-images.githubusercontent.com/7355199/81544694-22f40f00-9378-11ea-933b-d9a45e92ebfd.png">

HendrikThePendric commented:
I've had a quick attempt at reviewing these PRs but I got the distinct impression that the vast majority was just a matter of moving things around. @Mohammer5 would it be possible to indicate where you have actually changed an existing implementation?

Mohammer5 commented:
yeah, so most work has bee done on functions that use the value of the `selected` prop or a derived value of that prop. It's quite a few files as that was the biggest change because two things have changed at the same time:

1. Use strings instead of objects
2. Use `options` instead of `children` prop

If you want to, we could schedule a meeting where I go over what's changed @HendrikThePendric @ismay

HendrikThePendric commented on packages/widgets/src/Transfer/Transfer.js:
```diff
@@ -75,9 +64,9 @@ export const multiSelectedPropType = propTypes.arrayOf(singleSelectedPropType)
  * rest of the library
  */
 export const Transfer = ({
```
I've been thinking about rendering custom options and how we could stay a little bit more consistent here with other components that may require customisation. It's quite difficult to come up with something that is exactly the same, since this `options` prop pattern is fundamentally different than the `children` pattern we use elsewhere. After chatting a little bit with @ismay I came up with an idea (not sure whether Ismay will fully agree).

I'll place a few comments in the component code to list the changes I'd like to suggest.

HendrikThePendric commented on packages/widgets/src/Transfer/Transfer.js:
```diff
@@ -96,6 +85,7 @@ export const Transfer = ({
     leftFooter,
     leftHeader,
     maxSelections,
+    optionComponent: TransferOption,
```
change to an optional prop called `renderOption` of type function

HendrikThePendric commented on packages/widgets/src/Transfer/Transfer.js:
```diff
@@ -202,13 +201,28 @@ export const Transfer = ({
                 >
-                    {filteredSourceOptions}
+                    {sourceOptions.map(option => {
+                        const Option = option.component || TransferOption
+                        const highlighted = !!highlightedSourceOptions.find(
+                            highlightedSourceOption =>
+                                highlightedSourceOption === option.value
+                        )
+
+                        return (
+                            <Option
+                                key={option.value}
+                                disabled={option.disabled}
+                                option={option}
+                                highlighted={highlighted}
+                                {...getOptionClickHandlers(
+                                    option,
+                                    selectSingleOption,
+                                    toggleHighlightedSourceOption
+                                )}
+                            />
+                        )
```
Change this part to this
```jsx
const renderFn = option.render || renderOption
const optionProps = {
    disabled: option.disabled,
    option: option,
    highlighted: !!highlightedSourceOptions.find(
        highlightedSourceOption =>
            highlightedSourceOption === option.value,
        ...getOptionClickHandlers(
            option,
            selectSingleOption,
            toggleHighlightedSourceOption
        )
    ),
}

return typeof renderFn === 'function' ? (
    renderFn(optionProps)
) : (
    <TransferOption {...optionProps} />
)
```

HendrikThePendric commented on packages/widgets/src/Transfer/Transfer.js:
```diff
@@ -378,16 +399,23 @@ Transfer.defaultProps = {
+        propTypes.shape({
+            label: propTypes.string.isRequired,
+            value: propTypes.string.isRequired,
+            component: propTypes.func,
```
change to `render`

HendrikThePendric commented on packages/widgets/src/Transfer/Transfer.js:
```diff
@@ -401,15 +429,13 @@ Transfer.propTypes = {
     leftFooter: propTypes.node,
     leftHeader: propTypes.node,
     maxSelections: propTypes.oneOf([1, Infinity]),
+    optionComponent: propTypes.func,
```
change to `renderOption`

ismay commented:
@HendrikThePendric and I talked a bit about this. In general we're thinking of the same thing, though my preference would be a slightly more conservative approach. In pseudo code I'm thinking of:

```jsx
// Allow the user to override the default option rendering
renderOption = option.renderOption || defaultOptionRenderer

// option rendering just maps over all the options,
// and calls renderOption with the args needed for rendering options
// renderOption is expected to return (react) renderable content
options.map(option => {
  renderOption({ option, etc. })
})
```

So what I'm omitting compared to @HendrikThePendric's approach is allowing the render prop on an option in the options array. the way I see it, you can already implement everything you want in the `renderOption` callback, so I'd omit that for the sake of simplicity. We could always add it later on if it turns out to be necessary.

HendrikThePendric commented:
We also briefly discussed the option of adhering to the `children` API for v5, but not sure that is feasible at all.... Perhaps we can discuss things again tomorrow?

ismay commented:
> We also briefly discussed the option of adhering to the `children` API for v5, but not sure that is feasible at all.... Perhaps we can discuss things again tomorrow?

Yeah that'd be nice 👍. If we can align on everything tomorrow then we could maybe finish up the remaining tasks for v5 this week.

Mohammer5 commented:
I've introduced this the way @ismay proposed because there's a performance implication. The prop should either always expect a component or a function.

I've removed the custom option functionality from the individual option items as that can be done through the `renderOption` function. We can easily add it back if ever needed.
